finalfrontier/src/vocab.rs

```diff
@@ -376,16 +377,39 @@ where
     S: Hash + Eq + Clone + Ord,
 {
     fn from(builder: VocabBuilder<SimpleVocabConfig, T>) -> Self {
-        let min_count = builder.config.min_count;
+        let config = builder.config;
+        let vocab_cutoff = config.vocab_cutoff;
+        match vocab_cutoff {
+            VocabCutoff::MinCount(min_count) => {
+                let mut types: Vec<_> = builder
+                    .items
+                    .into_iter()
+                    .filter(|(_, count)| *count >= min_count as usize)
+                    .map(|(item, count)| CountedType::new(item.into(), count))
+                    .collect();
+                types.sort_unstable_by(|w1, w2| w2.cmp(&w1));
+                SimpleVocab::new(config, types, builder.n_items)
+            }
 
-        let mut types: Vec<_> = builder
-            .items
-            .into_iter()
-            .filter(|(_, count)| *count >= min_count as usize)
-            .map(|(item, count)| CountedType::new(item.into(), count))
-            .collect();
-        types.sort_unstable_by(|w1, w2| w2.cmp(&w1));
-        SimpleVocab::new(builder.config, types, builder.n_items)
+            VocabCutoff::TargetVocabSize(vocab_size) => {
+                assert!(vocab_size >= 1, "Target vocab size must be positive");
```

Comment:
I'm not sure whether we need this assertion. The value is guaranteed to be positive, because it's an unsigned integer. And in the case of someone entering `target_size=2`, I believe the training loop would never terminate because the negative sampling ensures not returning the same idx as the focus word as a negative sample in an unconditional loop.

Although having the assertion here doesn't hurt either.

Reply:
Nevermind, the `vocab_size - 1` below would underflow otherwise.